hoodie-client/src/main/java/com/uber/hoodie/io/HoodiePartitionHelper.java

```diff
@@ -0,0 +1,112 @@
+/*
+ * Copyright (c) 2017 Uber Technologies, Inc. (hoodie-dev-group@uber.com)
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *          http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.uber.hoodie.io;
+
+import com.uber.hoodie.common.model.HoodieRecord;
+import com.uber.hoodie.table.WorkloadProfile;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Set;
+import java.util.TreeSet;
+import scala.Tuple2;
+import scala.Tuple3;
+
+/**
+ * HoodiePartitionHelper is responsible for splitting Hoodie upsert stage into smaller spark
+ * stages.
+ */
+public class HoodiePartitionHelper {
+
+  /**
+   * Split an upsert workload
+   *
+   * @param profile WorkloadProfile for the input record
+   * @param isUpsert indicates upsert workflow or insert only workflow
+   * @param maxPartitions max spark partitions allowed for single datafile write stage
+   * @return List of conditions in Tuple3 of (lowBound, highBound, count)
+   */
+  public static List<Tuple3<String, String, Long>> splitWorkload(WorkloadProfile profile,
+      boolean isUpsert,
+      int maxPartitions) {
+
+    // We can't split input records if we don't know the workload or if this is a plain insert
+    if (profile == null || !isUpsert) {
+      return Collections.emptyList();
+    }
+
+    Set<String> sortedPartitions = new TreeSet<>();
+    List<Tuple2<String, Long>> splitPoints = new ArrayList<>();
+    sortedPartitions.addAll(profile.getPartitionPaths());
+    long totalPartitions = 0;
+    for (String partition : sortedPartitions) {
+      int updateCount = profile.getWorkloadStat(partition)
+          .getUpdateLocationToCount().size();
+      if (totalPartitions == 0
```

Comment:
Re-structure to : 

if(totalPartitions < maxPartitions) {
 totalPartitions += updateCount;
}
else {
 splitPoints.add(new Tuple2<>(partition, totalPartitions));
totalPartitions = updateCount;
}

Reply:
This would miss a case that if the 1 partition exceeds maxPartitions.  The code is put to ensure that at least 1 partition will be added to 1 split.